db: Latest Indexes

JediKev opened the pull request:
This adds all indexes we've implemented/accumulated over time.

aydreeihn committed:
Issue: Auto Referral

This commit fixes an issue we had with Auto Referrals. If a Ticket had been transfered/responded to by 3 or more departments and a user sent in an email response to one of the transferred departments (not including the original or current department), we created an auto referral for that department which could potentially give agents to access they should not have access to.

A new column, email_id, has been added to the thread_entry_email table so that we can see which system email each thread entry has been fetched from.

This commit also adds the event (EVENT_TABLE) relationship to the ThreadEvents (THREAD_EVENT_TABLE) class.

An index for ticket_pid on the ticket table has been added.

JediKev committed:
db: Latest Indexes

This adds all indexes we've implemented/accumulated over time.

JediKev commented on include/upgrader/streams/core/e7038ce9-fe7d449b.task.php:
```diff
@@ -0,0 +1,63 @@
+            list($t, $i) = explode(':', $index);
+
+            // Check if INDEX already exists via SHOW INDEX
+            $sql = sprintf("SHOW INDEX FROM `%s` WHERE `Key_name` = '%s'",
```
Column_name instead of Key_name

aydreeihn commented on setup/inc/streams/core/install-mysql.sql:
```diff
@@ -124,7 +124,8 @@ CREATE TABLE `%TABLE_PREFIX%form` (
     `updated` datetime NOT NULL,
-    PRIMARY KEY (`id`)
+    PRIMARY KEY (`id`),
+    INDEX `type` (`type`)
```
I think these are supposed to be called 'Key' rather than 'Index'.

JediKev committed:
forms: Task Title/Description Field Flags

This addresses an issue where the Task Title field allows any Admin to
remove the View and Required perms for Agents even though this field should
be Viewable and Required. This adds SQL so we can see if the
`FLAG_AGENT_REQUIRED`, `FLAG_AGENT_VIEW`, `FLAG_MASK_REQUIRE` and
`FLAG_MASK_VIEW` flags already exists, if not, we will add them to ensure
the field is Viewable and Required for Agents and to ensure Admins cannot
change this behavior for both perms. This also addresses an issue where the
Task Description field allows any Admin to remove the View perms for Agents
even though this field should always be Viewable. This adds SQL so we can
see if the `FLAG_AGENT_VIEW` and `FLAG_MASK_VIEW` flags already exists, if
not, we will add them to ensure the field is Viewable to Agents and to
ensure Admins cannot change this behavior.

protich commented on setup/inc/streams/core/install-mysql.sql:
```diff
@@ -820,7 +835,8 @@ CREATE TABLE IF NOT EXISTS `%TABLE_PREFIX%ticket_status` (
   UNIQUE KEY `name` (`name`),
-  KEY `state` (`state`)
+  KEY `state` (`state`),
+  KEY `id` (`id`)
```
`id` is already a PRIMARY KEY.

protich commented on setup/inc/streams/core/install-mysql.sql:
```diff
@@ -1014,7 +1033,10 @@ CREATE TABLE `%TABLE_PREFIX%user` (
-  KEY `org_id` (`org_id`)
+  KEY `org_id` (`org_id`),
+  KEY `default_email_id` (`default_email_id`),
+  KEY `id` (`id`),
```
We already have `PRIMARY KEY  (`id`),`

JediKev commented:
Updated

JediKev committed:
upgrade: Add Version Column To Plugin Table

This addresses an issue where we've never added the `version` column to any
of the upgrade patches so if someone were to upgrade from pre-1.8, they will
not have the `version` column and they will not be able to install plugins.
This adds a prepared statement to check if the column already exists, if it
does not exists we will add the column. We have to use a prepared statement
as unfortunately MySQL doesn’t have the concept of `ADD COLUMN IF NOT
EXISTS` like MariaDB does.

aydreeihn commented on include/upgrader/streams/core/e7038ce9-8cd26958.task.php:
```diff
@@ -0,0 +1,61 @@
+        'staff:isactive',
+        'staff:onvacation',
+        'task:flags',
+        'team_member:staff_id',
```
@JediKev I updated my PR to remove the Alter Table statement to add an index. So I'm thinking now you'd need rebase and add ticket:ticket_pid here, right?

JediKev replied:
@aydreeihn Alrighty rooster. Will do.

JediKev committed:
upgrader: Add ticket_pid Index

This updates the AddIndexMigration Task to include an index for the newly
added `ticket_pid` column in the Ticket table.

JediKev commented:
Updated

JediKev commented:
**Updated:** Includes 2FA template addition

protich commented on include/upgrader/streams/core/e7038ce9-8cd26958.patch.sql:
```diff
@@ -0,0 +1,47 @@
+/**
+ * @signature 8cd26958b50da5f037d722a172cb8e1f
+ * @version v1.14.2
```
not required but can you change this to `v1.15`
